File: keyval/class/class_test.go
package class

import (
	"testing"

	"github.com/stretchr/testify/require"

	q "github.com/janderland/fql/keyval"
)

func TestClassify(t *testing.T) {
	tests := []struct {
		kind Class
		kv   q.KeyValue
	}{
		{
			kind: Constant,
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("my"), q.String("dir")},
					Tuple:     q.Tuple{q.Int(123), q.Float(-55.8), q.String("wow")},
				},
				Value: q.Int(-38),
			},
		},
		{
			kind: VStampKey,
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("my"), q.String("dir")},
					Tuple:     q.Tuple{q.Int(123), q.VStampFuture{UserVersion: 202}, q.String("wow")},
				},
				Value: q.Int(-38),
			},
		},
		{
			kind: VStampVal,
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("my"), q.String("dir")},
					Tuple:     q.Tuple{q.Int(123), q.String("wow")},
				},
				Value: q.Tuple{q.VStampFuture{UserVersion: 202}},
			},
		},
		{
			kind: VStampVal,
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("my"), q.String("dir")},
					Tuple:     q.Tuple{q.Int(123), q.String("wow")},
				},
				Value: q.VStampFuture{UserVersion: 202},
			},
		},
		{
			kind: Clear,
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("my"), q.String("dir")},
					Tuple:     q.Tuple{q.Int(123), q.Float(-55.8), q.String("wow")},
				},
				Value: q.Clear{},
			},
		},
		{
			kind: ReadSingle,
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("my"), q.String("dir")},
					Tuple:     q.Tuple{q.Int(123), q.Float(-55.8), q.String("wow")},
				},
				Value: q.Variable{},
			},
		},
		{
			kind: ReadRange,
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.Variable{}, q.String("dir")},
					Tuple:     q.Tuple{q.Int(123), q.Float(-55.8), q.String("wow")},
				},
				Value: q.Int(-38),
			},
		},
		{
			kind: ReadRange,
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("my"), q.String("dir")},
					Tuple:     q.Tuple{q.Int(123), q.Variable{}, q.String("wow")},
				},
				Value: q.Nil{},
			},
		},
		{
			kind: ReadRange,
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.Variable{}, q.String("dir")},
					Tuple:     q.Tuple{q.Int(123), q.Float(-55.8), q.String("wow")},
				},
				Value: q.Variable{},
			},
		},
	}

	for _, test := range tests {
		t.Run(string(test.kind), func(t *testing.T) {
			kind := Classify(test.kv)
			require.Equal(t, test.kind, kind)
		})
	}
}

func TestClassifyInvalid(t *testing.T) {
	tests := []struct {
		name string
		kv   q.KeyValue
	}{
		{
			name: "nil directory",
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("hi"), nil, q.String("you")},
					Tuple:     nil,
				},
				Value: q.Bytes{},
			},
		},
		{
			name: "nil tuple",
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("hi"), q.String("you")},
					Tuple:     q.Tuple{q.Int(34), q.String("wow"), nil},
				},
				Value: q.Bytes{},
			},
		},
		{
			name: "nil value",
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("hi"), q.String("you")},
					Tuple:     q.Tuple{q.Int(34), q.String("wow")},
				},
				Value: nil,
			},
		},
		{
			name: "vstamp key & value",
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("where"), q.String("how?")},
					Tuple: q.Tuple{
						q.Int(42),
						q.VStampFuture{UserVersion: 384},
					},
				},
				Value: q.Tuple{q.VStampFuture{UserVersion: 384}},
			},
		},
		{
			name: "vstamp double",
			kv: q.KeyValue{
				Key: q.Key{
					Directory: q.Directory{q.String("where"), q.String("how?")},
					Tuple: q.Tuple{
						q.VStampFuture{UserVersion: 384},
						q.VStampFuture{UserVersion: 400},
					},
				},
				Value: q.Nil{},
			},
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			kind := Classify(test.kv)
			require.Regexp(t, "invalid", string(kind))
		})
	}
}

File: elixir/fql/test/fql/keyval/class_test.exs
defmodule Fql.KeyVal.ClassTest do
  use ExUnit.Case
  alias Fql.KeyVal
  alias Fql.KeyVal.Class

  describe "classify/1" do
    test "classifies constant query" do
      kv = KeyVal.key_value(
        KeyVal.key(["dir"], [KeyVal.int(1)]),
        "value"
      )

      assert Class.classify(kv) == :constant
    end

    test "classifies clear query" do
      kv = KeyVal.key_value(
        KeyVal.key(["dir"], [KeyVal.int(1)]),
        :clear
      )

      assert Class.classify(kv) == :clear
    end

    test "classifies read_single query" do
      kv = KeyVal.key_value(
        KeyVal.key(["dir"], [KeyVal.int(1)]),
        KeyVal.variable([:string])
      )

      assert Class.classify(kv) == :read_single
    end

    test "classifies read_range query with variable in key" do
      kv = KeyVal.key_value(
        KeyVal.key(["dir"], [KeyVal.variable([:int])]),
        KeyVal.variable([])
      )

      assert Class.classify(kv) == :read_range
    end

    test "classifies vstamp_key query" do
      kv = KeyVal.key_value(
        KeyVal.key(["dir"], [KeyVal.vstamp_future(0)]),
        "value"
      )

      assert Class.classify(kv) == :vstamp_key
    end

    test "classifies vstamp_val query" do
      kv = KeyVal.key_value(
        KeyVal.key(["dir"], [KeyVal.int(1)]),
        KeyVal.vstamp_future(0)
      )

      assert Class.classify(kv) == :vstamp_val
    end

    test "classifies invalid query with nil" do
      kv = KeyVal.key_value(
        KeyVal.key(["dir"], [:nil]),
        "value"
      )

      assert {:invalid, _} = Class.classify(kv)
    end

    test "classifies invalid query with multiple vstamp futures" do
      kv = KeyVal.key_value(
        KeyVal.key(["dir"], [KeyVal.vstamp_future(0)]),
        KeyVal.vstamp_future(0)
      )

      assert {:invalid, _} = Class.classify(kv)
    end
  end
end
